examples/nlp/text_generation_with_miniature_gpt.py

```diff
@@ -0,0 +1,331 @@
+        )  # (batch_size, seq_len, embed_dim)
+        output = self.combine_heads(
+            concat_attention
+        )  # (batch_size, seq_len, embed_dim)
+        return output
+
+
+"""
+## Implement a Transformer block as a layer
+"""
+
+
+class TransformerBlock(layers.Layer):
+    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
+        super(TransformerBlock, self).__init__()
+        self.att = MultiHeadSelfAttention(embed_dim, num_heads)
+        self.ffn = keras.Sequential(
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
+        )
+        self.layernorm1 = layers.LayerNormalization(epsilon=1e-6)
+        self.layernorm2 = layers.LayerNormalization(epsilon=1e-6)
+        self.dropout1 = layers.Dropout(rate)
+        self.dropout2 = layers.Dropout(rate)
+
+    def call(self, inputs, training):
+        attn_output = self.att(inputs)
+        attn_output = self.dropout1(attn_output, training=training)
+        out1 = self.layernorm1(inputs + attn_output)
+        ffn_output = self.ffn(out1)
+        ffn_output = self.dropout2(ffn_output, training=training)
+        return self.layernorm2(out1 + ffn_output)
+
+
+"""
+## Implement embedding layer
+
+Two seperate embedding layers, one for tokens, one for token index (positions).
+"""
+
+
+class TokenAndPositionEmbedding(layers.Layer):
+    def __init__(self, maxlen, vocab_size, emded_dim):
+        super(TokenAndPositionEmbedding, self).__init__()
+        self.token_emb = layers.Embedding(input_dim=vocab_size, output_dim=emded_dim)
+        self.pos_emb = layers.Embedding(input_dim=maxlen, output_dim=emded_dim)
+
+    def call(self, x):
+        maxlen = tf.shape(x)[-1]
+        positions = tf.range(start=0, limit=maxlen, delta=1)
+        positions = self.pos_emb(positions)
+        x = self.token_emb(x)
+        return x + positions
+
+
+"""
+## Implement miniature GPT model
+"""
+vocab_size = 20000  # Only consider the top 20k words
+maxlen = 200  # Max sequence size
+embed_dim = 32  # Embedding size for each token
+num_heads = 2  # Number of attention heads
+ff_dim = 32  # Hidden layer size in feed forward network inside transformer
+
+
+def create_model():
+    inputs = layers.Input(shape=(maxlen,))
+    embedding_layer = TokenAndPositionEmbedding(maxlen, vocab_size, embed_dim)
+    x = embedding_layer(inputs)
+    transformer_block = TransformerBlock(embed_dim, num_heads, ff_dim)
+    x = transformer_block(x)
+    outputs = layers.Dense(vocab_size, activation="softmax")(x)
+    model = keras.Model(inputs=inputs, outputs=[outputs, x])
+    model.compile(
+        "adam", loss=["sparse_categorical_crossentropy", None],
+    )
+    return model
+
+
+"""
+## Prepare data for word level language modelling
+
+We will download IMDB data, and combine training and validation sets for
+text generation task.
+"""
+
+
+(x_train, y_train), (x_val, y_val) = keras.datasets.imdb.load_data(num_words=vocab_size)
```

Comment:
Consider starting from the raw data and using the `TextVectorization` layer for tokenization (it already handles OOV and padding). The shifting can be done via `dataset.map()`. This would make the example much more generalizable since people could drop in their own folder of texts.

Ref: https://keras.io/examples/nlp/text_classification_from_scratch/

Reply:
Just noticed, the referred example doesn't work in google colab unless you do a `!pip install tf-nightly` and restart the runtime. Maybe there should be a warning line in that example till `text_dataset_from_directory` becomes a part of the stable release.

Reply:
Yes, in general everything on keras.io expects tf-nightly at this time. These features will be in TF 2.3, which will be out within a month or so. So this issue is temporary.

Reply:
And Colab TPU throws an error with `tf-nightly`.

Reply:
Please file a report, and try to install an earlier version of tf-nightly: https://pypi.org/project/tf-nightly/#history